VACMS-1872: Render child sections.

oksana-c opened the pull request:
## Description

See #1872 . 

## Testing done


## Screenshots

### Access to All Sections
<img width="1280" alt="Screen Shot 2020-06-05 at 7 49 03 AM" src="https://user-images.githubusercontent.com/16477724/83883677-0b9f0c00-a701-11ea-9423-2e2f52cfefda.png">

### Access to some sections

<img width="1110" alt="Screen Shot 2020-06-05 at 7 53 17 AM" src="https://user-images.githubusercontent.com/16477724/83884103-a3045f00-a701-11ea-9d69-67ff2ea3dde1.png">


## QA steps

As user _uid_ with _user_role_
1. Do this
1. Then that
1. Then validate Acceptance Criteria from issue
- [ ] This
- [ ] That
- [ ] The other thing

## Definition of Done
- [ ] Product release notes 
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

swirtSJW commented:
508 Issues:  Needs to use aria to reflect its state.  Also needs to work on keyboard navigation similar to 
https://designsystem.digital.gov/components/accordion/
You could look at node_link_report to see how it is handled.
See this page to test keyboard  and aria handling
http://pr1980.ci.cms.va.gov/locations/franklin-r-sousley-campus

swirtSJW commented:
The other change I noticed is that I no longer have sections.
![image](https://user-images.githubusercontent.com/5752113/83895714-16699900-a721-11ea-9c14-3aa5b99f9b1e.png)

oksana-c commented:
@todo: add user path to accessibility testiny

oksana-c commented:
alrighty! switched this baby to use accordion pattern from VADS and added other style adjustments. It is now 100% compliant and we won't need another iteration just to cover 508.

Users with "bypass workbench access" permission will see all sections.

<img width="481" alt="Screen Shot 2020-06-05 at 2 26 48 PM" src="https://user-images.githubusercontent.com/16477724/83920678-f85c6280-a739-11ea-9269-c571cf2cccd3.png">

also, added /user path to accessibility check. Let's get ready for some warnings 😄 :octocat:

oksana-c commented:
@kevwalsh ^^ fyi

oksana-c commented:
Another update in this PR that is worth noting - **\Drupal\va_gov_user\Service\UserPermsService::getSections** method was refactored to return an array of all sections that user has access to, keyed by TIDs. 

This update removes tight coupling to Workbench Sections key `administration`, which (if returned by getSections, signified that the user has access to _all sections_.  Removing the dependency on this key will allows us to rely on specific tids and prevent us from adding any proprietary logic (tech-debt) which function would rely on workbench-specific definitions.

---

/user path that I added to accessibility tests in a previous commit shows 1 contrast error. Detected contrast error is coming from the administrator toolbar and is unrelated to the "sections" block added in this PR. We have #1302 open to fix accessibility tests that are global to the application.

swirtSJW commented:
@oksana-c I like the accessibility improvements and the switch to full accordions.   
One minor concern is the pattern of putting a  link in the button. area.

It is not accounted for on either 
https://designsystem.digital.gov/components/accordion/
or 
https://design.va.gov/components/accordions

In this case it creates some additional 508 issues.
1.  On tabbing the focus seems to go from right to left as it jumps from the plus sign to the link texts. 
 This is counter to movement of the eye and reading direction.
2.  Since there is no text in the button itself, the screen reader just reads "Toggle",   and then reads the section link, but by that time the reader  is on the section link, so by the time they are informed what the toggle does, they are on the section link which takes them to the section.
I think you have to  put the heading in the button, without the link, then repeat the list item within the link inside the accordion content.
3.  The shading behavior on hover.   There are two hover zones that carry out actions.  The link text and the plus, but there is also a hover zone in the middle that responds as though it should do something, but does nothing.  Addressing issue number 2 would also resolve this issue.
![image](https://user-images.githubusercontent.com/5752113/84044369-7789aa80-a975-11ea-8bb2-ad9f54a1eec3.png)

If we want to get this out, as-is, and target this issue with a follow-up, I am okay with that as long as we commit to fixing it.    @kevwalsh Any thoughts on timing?

oksana-c commented:
after thorough deliberation, decision was made to use "menu" pattern for Sections. Screenshots for the most common scenarios of section assignment are attached below

<img width="1117" alt="Screen Shot 2020-06-08 at 12 04 36 PM" src="https://user-images.githubusercontent.com/16477724/84064757-6d20de00-a980-11ea-98d8-b84d7d164700.png">

<img width="1099" alt="Screen Shot 2020-06-08 at 12 06 24 PM" src="https://user-images.githubusercontent.com/16477724/84064862-9a6d8c00-a980-11ea-8710-24143ab37141.png">

<img width="1117" alt="Screen Shot 2020-06-08 at 12 06 37 PM" src="https://user-images.githubusercontent.com/16477724/84064869-9d687c80-a980-11ea-827a-59c6bb298f28.png">

oksana-c commented:
@swirtSJW ,  Give this another whirl once devshop env builds.

@kevwalsh is onboard w/ the latest "menu" pattern and the display of child sections as currently implemented (i.e. not nested). Our goal is to reduce the number of clicks the user takes to get to the content of any section. By allowing all child sections to be visible without extra collapsible areas, we are moving closer to that goal.

Please test keyboard navigation between sections while paying special attention to the order in which parent section link and collapse button are tabbed. Press "return" while focused on the collapse button and tab through child sections.

P.S. I had to switch parent link render array to markup element instead o f a link. I am not fond of that solution, but there's no other way that I know of to achieve the order we need w/ render arrays. We aim for perfect, knowing it unattainable...

oksana-c commented:
closing the loop here:
> There is a bit of mismatch between hover target and link target. but it is a minor issue.

current implementation follows core menu pattern styles for link targets. We are releasing as-is with an understanding that further iterations may follow with a potential pattern change.
